File: Sources/OCRAdminCommands.swift
import Foundation

enum OCRAdminCommands {
  static let supportedCommands: Set<String> = [
    "index-existing",
    "rebuild-index",
    "index-status",
    "clear-index",
  ]

  static func runIfNeeded(arguments: [String] = CommandLine.arguments) -> Bool {
    guard arguments.count >= 2 else { return false }
    let command = arguments[1]
    guard supportedCommands.contains(command) else { return false }

    switch command {
    case "index-existing":
      let summary = OCRIndexManager.shared.runBatchSync(mode: .indexExisting)
      printSummary("Index existing complete", summary)
    case "rebuild-index":
      let summary = OCRIndexManager.shared.runBatchSync(mode: .rebuild)
      printSummary("Rebuild complete", summary)
    case "index-status":
      printStatus()
    case "clear-index":
      OCRIndexManager.shared.clearIndex()
      print("OCR index cleared.")
    default:
      return false
    }
    return true
  }

  private static func printSummary(_ title: String, _ summary: OCRBatchSummary) {
    print(title)
    print("total: \(summary.total)")
    print("indexed: \(summary.indexed)")
    print("skipped: \(summary.skipped)")
    print("missing: \(summary.missing)")
    print("failed: \(summary.failed)")
    if summary.cancelled {
      print("cancelled: true")
    }
  }

  private static func printStatus() {
    let records = UploadHistoryStore.shared.snapshot().filter { $0.isImageRecord }
    var counts: [String: Int] = [:]
    for record in records {
      let key = record.ocrStatus?.rawValue ?? "notIndexed"
      counts[key, default: 0] += 1
    }

    print("OCR index status")
    print("imageRecords: \(records.count)")
    for key in ["indexed", "pending", "failed", "missingFile", "disabled", "skipped", "notIndexed"] {
      if let value = counts[key] {
        print("\(key): \(value)")
      }
    }
    print("enabled: \(RuntimePreferences.shared.ocrIndexingEnabled)")
  }
}

File: Windows/CraftyCannon.Ocr/OcrAdminCommands.cs
using CraftyCannon.Core;

namespace CraftyCannon.Ocr;

public static class OcrAdminCommands
{
    private static readonly HashSet<string> SupportedCommands = new(StringComparer.Ordinal)
    {
        "index-existing",
        "rebuild-index",
        "index-status",
        "clear-index"
    };

    public static bool IsSupported(IReadOnlyList<string> arguments) =>
        arguments.Count >= 1 && SupportedCommands.Contains(arguments[0]);

    public static async Task<bool> RunIfNeededAsync(
        IReadOnlyList<string> arguments,
        IOcrIndexingService indexer,
        IUploadHistoryStore history,
        bool enabled,
        TextWriter output,
        CancellationToken cancellationToken = default)
    {
        if (!IsSupported(arguments))
        {
            return false;
        }

        switch (arguments[0])
        {
            case "index-existing":
                PrintSummary("Index existing complete", await indexer.RunBatchAsync(OcrBatchMode.IndexExisting, cancellationToken).ConfigureAwait(false), output);
                return true;
            case "rebuild-index":
                PrintSummary("Rebuild complete", await indexer.RunBatchAsync(OcrBatchMode.Rebuild, cancellationToken).ConfigureAwait(false), output);
                return true;
            case "index-status":
                PrintStatus(history, enabled, output);
                return true;
            case "clear-index":
                await indexer.ClearIndexAsync(cancellationToken).ConfigureAwait(false);
                output.WriteLine("OCR index cleared.");
                return true;
            default:
                return false;
        }
    }

    private static void PrintSummary(string title, OcrBatchSummary summary, TextWriter output)
    {
        output.WriteLine(title);
        output.WriteLine($"total: {summary.Total}");
        output.WriteLine($"indexed: {summary.Indexed}");
        output.WriteLine($"skipped: {summary.Skipped}");
        output.WriteLine($"missing: {summary.Missing}");
        output.WriteLine($"failed: {summary.Failed}");
    }

    private static void PrintStatus(IUploadHistoryStore history, bool enabled, TextWriter output)
    {
        var records = history.Records.Where(OcrIndexingService.IsImageRecord).ToArray();
        var counts = records
            .GroupBy(record => StatusKey(record.OcrStatus))
            .ToDictionary(group => group.Key, group => group.Count(), StringComparer.Ordinal);

        output.WriteLine("OCR index status");
        output.WriteLine($"imageRecords: {records.Length}");
        foreach (var key in new[] { "indexed", "pending", "failed", "missingFile", "disabled", "skipped", "notIndexed" })
        {
            if (counts.TryGetValue(key, out var count))
            {
                output.WriteLine($"{key}: {count}");
            }
        }

        output.WriteLine($"enabled: {enabled.ToString().ToLowerInvariant()}");
    }

    private static string StatusKey(OcrIndexStatus status) => status switch
    {
        OcrIndexStatus.Indexed => "indexed",
        OcrIndexStatus.Pending => "pending",
        OcrIndexStatus.Failed => "failed",
        OcrIndexStatus.LocalImageMissing => "missingFile",
        OcrIndexStatus.Disabled => "disabled",
        OcrIndexStatus.Skipped => "skipped",
        _ => "notIndexed"
    };
}
